src/input/sources/hls.rs

```diff
@@ -0,0 +1,139 @@
+use std::{
+    io::{ErrorKind as IoErrorKind, Result as IoResult, SeekFrom},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use futures::StreamExt;
+use pin_project::pin_project;
+use reqwest::{header::HeaderMap, Client};
+use serenity::async_trait;
+use stream_lib::{DownloadStream, Event};
+use symphonia_core::io::MediaSource;
+use tokio::io::{AsyncRead, AsyncSeek, ReadBuf};
+use tokio_util::io::StreamReader;
+
+use crate::input::{
+    AsyncAdapterStream,
+    AsyncMediaSource,
+    AudioStream,
+    AudioStreamError,
+    Compose,
+    Input,
+};
+
+/// Lazy HLS stream
+#[derive(Debug)]
+pub struct HlsRequest {
+    /// HLS downloader
+    pub hls: Option<DownloadStream>,
+}
+
+impl HlsRequest {
+    #[must_use]
+    /// Create a lazy HLS request.
+    pub fn new(client: Client, request: &str) -> Self {
+        Self::new_with_headers(client, request, HeaderMap::default())
+    }
+
+    #[must_use]
+    /// Create a lazy HTTP request.
+    pub fn new_with_headers(client: Client, request: &str, headers: HeaderMap) -> Self {
+        let request = client.get(request).headers(headers).build().unwrap();
+        let hls = stream_lib::download_hls(client, request, None);
+
+        HlsRequest { hls: Some(hls) }
+    }
+
+    fn create_stream(&mut self) -> Result<HlsStream, AudioStreamError> {
+        let Some(hls) = self.hls.take() else {
+            return Err(AudioStreamError::Fail("hls can only be used once".into()));
+        };
+
+        let stream = Box::new(StreamReader::new(hls.map(|ev| match ev {
+            Event::Bytes { bytes } => Ok(bytes),
```

Comment:
I guess a question on how `stream_lib` works: is it downloading all chunks in sequence in the background and holding onto them, or are you doing something like a fixed-size lookahead? I'm really just curious here, since the cost of holding an Opus-encoded song in memory is like <10MB which I'd consider acceptable for enabling HLS support.

Reply:
So at the moment it just downloads all chunks in the hls playlist and queue them in, as they are put into bytes they are reference counted so I think they should be dropped as they are played.

I will have to think about it a bit to figure out how to do a fixed size lookahead (either time or size based) but it is something that could be added.

The place where it happens: https://github.com/Erk-/rsget/blob/master/stream_lib/src/hls/mod.rs#L117

(The code could use a clean up, looking at that file now I can see it is quite messy)